File: polyA/load_alignments.py
import re
from typing import Iterable, List, NamedTuple, Optional, TextIO, Tuple

from .exceptions import FileFormatException
from .alignment import Alignment, get_skip_state
from .constants import INFINITE_SHARD_GAP
from .performance import timeit


def _parse_meta_line(line: str) -> Optional[Tuple[str, str]]:
    """
    Attempt to parse one line of a Stockholm file. If the line
    contains metadata, return a tuple of the key and value.
    Otherwise, return ``None``.

    >>> _parse_meta_line("#=GF ID\tSeqMcSeq")
    ('ID', 'SeqMcSeq')
    >>> _parse_meta_line("seqname\tatcg")
    >>> _parse_meta_line("//")
    >>> _parse_meta_line("# stockholm 1.0")
    """
    if line.strip().startswith("#="):
        parts = line.strip().split()

        if len(parts) != 3:
            return None

        return parts[1], parts[2]

    return None


def _parse_alignment_line(line: str) -> Optional[Tuple[str, str]]:
    """
    Attempt to parse one line of a sequence alignment,
    returning a tuple of the sequence name and the sequence
    itself. If the string isn't formatted correctly, return
    ``None``.

    >>> _parse_alignment_line("seqname\tatcg")
    ('seqname', 'atcg')
    >>> _parse_alignment_line("#=GF ID SeqMcSeq")
    >>> _parse_meta_line("//")
    >>> _parse_meta_line("# stockholm 1.0")
    """
    parts = line.strip().split()

    if len(parts) != 2:
        return None

    return parts[0], parts[1]


def _parse_preamble_line(line: str) -> bool:
    """
    Return ``True`` if this line contains a Stockholm format
    preamble ("# Stockholm 1.0") and ``False`` otherwise.
    """
    return line.strip().upper().startswith("# STOCKHOLM")


def _parse_tool_line(line: str) -> str:
    """
    Return ``True`` if this line contains a tool preamble of
    the format we add in the converter scripts.

    >>> _parse_tool_line("# ALIGNMENT TOOL cross_match")
    'cross_match'
    >>> _parse_tool_line("# Something else")
    ''
    """
    if line.strip().upper().startswith("# ALIGNMENT TOOL"):
        return " ".join(line.strip().split()[3:])
    else:
        return ""


def _parse_terminator_line(line: str) -> bool:
    """
    Return ``True`` if this line contains a terminator
    ("//") or ``False`` otherwise.

    >>> _parse_terminator_line(" // ")
    True
    >>> _parse_terminator_line(" / / ")
    False
    """
    return line.strip() == "//"


def _parse_chrom_meta(line: str) -> Optional[Tuple[str, int, int]]:
    match = re.search(r"(.+):(\d+)-(\d+)", line.strip())
    if match is None:
        return None

    chrom_name = match.groups()[0]
    chrom_start = int(match.groups()[1])
    chrom_stop = int(match.groups()[2])

    return chrom_name, chrom_start, chrom_stop


@timeit()
def load_alignments(
    file: TextIO, add_skip_state: bool = False
) -> Iterable[Alignment]:
    """
    Load a set of alignments in Stockholm format.
    See below for an example.

    ::

        #=GF ID  MERX#DNA/TcMar-Tigger
        #=GF TR  chr0:0000-0000
        #=GF SC  1153
        #=GF ST  +
        #=GF TQ  -1
        #=GF ST  127
        #=GF SP  601
        #=GF CST 135
        #=GF CSP 628
        #=GF FL  128
        #=GF MX  20p41g.matrix
        #=GF GI -25
        #=GF GE -5
    """
    if add_skip_state:
        yield get_skip_state()

    meta = {}
    seqs = []

    current_line_number = 1

    for line in file:
        if _parse_preamble_line(line):
            continue

        if _parse_tool_line(line):
            continue

        parts = _parse_meta_line(line)
        if parts is not None:
            meta[parts[0]] = parts[1]
            continue

        parts = _parse_alignment_line(line)
        if parts is not None:
            seqs.append(parts[1])
            continue

        if _parse_terminator_line(line):
            if len(seqs) != 2:
                raise FileFormatException(
                    file.name,
                    current_line_number,
                    "wrong number of sequences",
                )

            try:
                chrom_meta = _parse_chrom_meta(meta["TR"])
                if chrom_meta is None:
                    raise FileFormatException(
                        file.name,
                        current_line_number,
                        "missing TR field",
                    )

                chrom_name, chrom_start, chrom_stop = chrom_meta

                if meta["TQ"] == "t":
                    yield Alignment(
                        subfamily=meta["ID"],
                        chrom_name=chrom_name,
                        chrom_start=chrom_start,
                        chrom_stop=chrom_stop,
                        score=int(meta["SC"]),
                        start=int(meta["ST"]),
                        stop=int(meta["SP"]),
                        consensus_start=int(meta["CST"]),
                        consensus_stop=int(meta["CSP"]),
                        sequences=[seqs[0][::-1], seqs[1][::-1]],
                        strand=meta["SD"],
                        flank=int(meta["FL"]),
                        sub_matrix_name=meta["MX"],
                        gap_init=float(meta["GI"]),
                        gap_ext=float(meta["GE"]),
                    )
                else:
                    yield Alignment(
                        subfamily=meta["ID"],
                        chrom_name=chrom_name,
                        chrom_start=chrom_start,
                        chrom_stop=chrom_stop,
                        score=int(meta["SC"]),
                        start=int(meta["ST"]),
                        stop=int(meta["SP"]),
                        consensus_start=int(meta["CST"]),
                        consensus_stop=int(meta["CSP"]),
                        sequences=[seqs[0], seqs[1]],
                        strand=meta["SD"],
                        flank=int(meta["FL"]),
                        sub_matrix_name=meta["MX"],
                        gap_init=float(meta["GI"]),
                        gap_ext=float(meta["GE"]),
                    )
            except IndexError as e:
                raise FileFormatException(
                    file.name,
                    current_line_number,
                    "invalid TR field",
                )
            except KeyError as e:
                raise FileFormatException(
                    file.name,
                    current_line_number,
                    f"missing key: {e.args[0]}",
                )
            except ValueError as e:
                raise FileFormatException(
                    file.name,
                    current_line_number,
                    e.args[0],
                )

            meta.clear()
            seqs.clear()

        current_line_number += 1


def load_alignment_tool(file: TextIO) -> str:
    """
    Return tool used to produce the alignments in the input file.
    """

    while True:
        try:
            # skip header line
            header_line = next(file)
        except StopIteration:
            return ""

        try:
            line = next(file)
            tool = _parse_tool_line(line)

            return tool if tool else ""
        except StopIteration:
            return ""


class Shard(NamedTuple):
    start: int
    stop: int
    alignments: List[Alignment]


def shard_overlapping_alignments(
    alignments: Iterable[Alignment],
    shard_gap: int,
    add_skip_state: bool = True,
) -> Iterable[Shard]:
    """
    Shard the given alignments into overlapping groups, separated by no more
    than `shard_gap` nucleotides. This allows for more efficient processing
    for alignments over large regions of sequence (such as an entire genome)
    where many regions will be empty.
    Precondition: alignments are sorted by their start position and all
    alignments have start position <= stop position.

    >>> skip = get_skip_state()
    >>> a0 = Alignment("", "a", 1, 100, 0, 1, 10, 0, 0, [], "", 0, "", 0, 0)
    >>> a1 = Alignment("", "a", 1, 100, 0, 21, 30, 0, 0, [], "", 0, "", 0, 0)
    >>> shards = list(shard_overlapping_alignments([a0, a1], 10))
    >>> len(shards)
    2
    >>> shards[0].start
    1
    >>> shards[0].stop
    15
    >>> len(shards[0].alignments)
    2
    >>> shards[1].start
    16
    >>> shards[1].stop
    100
    >>> len(shards[1].alignments)
    2
    """
    shard_alignments: List[Alignment] = (
        [get_skip_state()] if add_skip_state else []
    )

    shard_start = 1
    shard_stop = None

    is_infinite_gap = shard_gap == INFINITE_SHARD_GAP

    last_alignment = None

    for alignment in alignments:
        last_alignment = alignment

        # If this is the first alignment we need to initialize
        # the stop position
        if shard_stop is None:
            shard_stop = alignment.stop

        if is_infinite_gap or alignment.start <= (shard_stop + shard_gap):
            # This alignment might extend past the previous
            # alignments in the shard, capture that here
            if shard_stop < alignment.stop:
                shard_stop = alignment.stop

            shard_alignments.append(alignment)
        else:
            # The gap between shards might be bigger than the
            # maximum shard gap, so we have to consume half of
            # the actual gap for this shard
            actual_gap = alignment.start - shard_stop
            shard_stop += int(actual_gap / 2)

            yield Shard(
                start=shard_start,
                stop=shard_stop,
                alignments=shard_alignments,
            )

            shard_start = shard_stop + 1
            shard_stop = alignment.stop

            # Note: important to create a new list here or we will
            # mutate the one we just handed back to the caller.
            shard_alignments = (
                [get_skip_state(), alignment] if add_skip_state else [alignment]
            )

    if last_alignment is None:
        raise ValueError("no alignments found")

    yield Shard(
        start=shard_start,
        stop=last_alignment.chrom_stop - last_alignment.chrom_start + 1,
        alignments=shard_alignments,
    )

File: load_alignments.go
package polyA

import (
	"bufio"
	"io"
	"regexp"
	"strconv"
	"strings"
)

var chromMetaRegexp *regexp.Regexp

func init() {
	chromMetaRegexp = regexp.MustCompile("(.+):(\\d+)-(\\d+)")
}

func parseMetaLine(line string) (key string, value string, found bool) {
	trimmedLine := strings.Trim(line, " \t")
	if strings.HasPrefix(trimmedLine, "#=") {
		parts := strings.Fields(trimmedLine)
		if len(parts) != 3 {
			return "", "", false
		}

		return parts[1], parts[2], true
	}

	return "", "", false
}

func parseAlignmentLine(line string) (key string, value string, found bool) {
	trimmedLine := strings.Trim(line, " \t")
	parts := strings.Fields(trimmedLine)
	if len(parts) != 2 {
		return "", "", false
	}

	return parts[0], parts[1], true
}

func parsePreambleLine(line string) bool {
	trimmedLine := strings.Trim(line, " \t")
	upperLine := strings.ToUpper(trimmedLine)
	return strings.HasPrefix(upperLine, "# STOCKHOLM")
}

func parseToolLine(line string) string {
	trimmedLine := strings.Trim(line, " \t")
	upperLine := strings.ToUpper(trimmedLine)
	if strings.HasPrefix(upperLine, "# ALIGNMENT TOOL") {
		parts := strings.Fields(trimmedLine)
		return strings.Join(parts[3:], " ")
	}

	return ""
}

func parseTerminatorLine(line string) bool {
	return strings.Trim(line, " \t") == "//"
}

func parseChromMeta(line string) (name string, start int, stop int, found bool) {
	trimmedLine := strings.Trim(line, " \t")
	match := chromMetaRegexp.FindStringSubmatch(trimmedLine)
	if match == nil {
		return "", 0, 0, false
	}

	chromName := match[1]
	chromStart, err := strconv.Atoi(match[2])
	if err != nil {
		panic("bad match - failed to convert to int")
	}
	chromStop, err := strconv.Atoi(match[3])
	if err != nil {
		panic("bad match - failed to convert to int")
	}

	return chromName, chromStart, chromStop, true
}

type Loader struct {
	addSkipState bool
	done         bool
	loadedCount  int
	scanner      *bufio.Scanner
}

func NewLoader(file io.Reader, addSkipState bool) *Loader {
	return &Loader{
		addSkipState: addSkipState,
		scanner:      bufio.NewScanner(file),
	}
}

func (l *Loader) Load(a *Alignment) bool {
	if l.done {
		return false
	}

	if l.loadedCount == 0 && l.addSkipState {
		a.Subfamily = skipState.Subfamily
		a.ChromName = skipState.ChromName
		a.ChromStart = skipState.ChromStart
		a.ChromStop = skipState.ChromStop
		a.Score = skipState.Score
		a.Start = skipState.Start
		a.Stop = skipState.Stop
		a.ConsensusStart = skipState.ConsensusStart
		a.ConsensusStop = skipState.ConsensusStop
		a.Sequence = skipState.Sequence
		a.SubfamilySequence = skipState.SubfamilySequence
		a.Strand = skipState.Strand
		a.Flank = skipState.Flank
		a.SubMatrixName = skipState.SubMatrixName
		a.GapInit = skipState.GapInit
		a.GapExt = skipState.GapExt

		l.loadedCount += 1
		return true
	}

	meta := make(map[string]string)
	var seqs []string

	for {
		more := l.scanner.Scan()
		line := l.scanner.Text()

		if parsePreambleLine(line) {
			continue
		}

		if parseToolLine(line) != "" {
			continue
		}

		key, value, found := parseMetaLine(line)
		if found {
			meta[key] = value
			continue
		}

		_, seq, found := parseAlignmentLine(line)
		if found {
			seqs = append(seqs, seq)
			continue
		}

		if parseTerminatorLine(line) {
			if len(seqs) != 2 {
				panic("file format error - wrong number of sequences")
			}

			chromName, chromStart, chromStop, found := parseChromMeta(meta["TR"])
			if !found {
				panic("file format error - missing TR field")
			}

			// TODO: Write helper(s) to check for presence of key and then get the value

			score, err := strconv.Atoi(meta["SC"])
			if err != nil {
				panic("file format error - invalid SC field")
			}

			start, err := strconv.Atoi(meta["ST"])
			if err != nil {
				panic("file format error - invalid ST field")
			}

			stop, err := strconv.Atoi(meta["SP"])
			if err != nil {
				panic("file format error - invalid SP field")
			}

			consensusStart, err := strconv.Atoi(meta["CST"])
			if err != nil {
				panic("file format error - invalid CST field")
			}

			consensusStop, err := strconv.Atoi(meta["CSP"])
			if err != nil {
				panic("file format error - invalid CSP field")
			}

			var sequence string
			var subfamilySequence string
			if meta["TQ"] == "t" {
				sequence = reverse(seqs[0])
				subfamilySequence = reverse(seqs[1])
			} else {
				sequence = seqs[0]
				subfamilySequence = seqs[1]
			}

			flank, err := strconv.Atoi(meta["FL"])
			if err != nil {
				panic("file format error - invalid FL field")
			}

			gapInit, err := strconv.ParseFloat(meta["GI"], 64)
			if err != nil {
				panic("file format error - invalid GI field")
			}

			gapExt, err := strconv.ParseFloat(meta["GE"], 64)
			if err != nil {
				panic("file format error - invalid GE field")
			}

			a.Subfamily = meta["ID"]
			a.ChromName = chromName
			a.ChromStart = chromStart
			a.ChromStop = chromStop
			a.Score = score
			a.Start = start
			a.Stop = stop
			a.ConsensusStart = consensusStart
			a.ConsensusStop = consensusStop
			a.Sequence = sequence
			a.SubfamilySequence = subfamilySequence
			a.Strand = meta["SD"]
			a.Flank = flank
			a.SubMatrixName = meta["MX"]
			a.GapInit = gapInit
			a.GapExt = gapExt

			l.loadedCount += 1

			if !more {
				l.done = true
				l.scanner = nil
			}

			return true
		}

		if !more {
			l.done = true
			l.scanner = nil

			return false
		}
	}
}

func LoadAlignments(file io.Reader, addSkipState bool) <-chan *Alignment {
	loaded := make(chan *Alignment)

	go func() {
		if addSkipState {
			loaded <- SkipState()
		}

		meta := make(map[string]string)
		var seqs []string

		scanner := bufio.NewScanner(file)
		for {
			more := scanner.Scan()
			line := scanner.Text()

			if parsePreambleLine(line) {
				continue
			}

			if parseToolLine(line) != "" {
				continue
			}

			key, value, found := parseMetaLine(line)
			if found {
				meta[key] = value
				continue
			}

			_, seq, found := parseAlignmentLine(line)
			if found {
				seqs = append(seqs, seq)
				continue
			}

			if parseTerminatorLine(line) {
				if len(seqs) != 2 {
					panic("file format error - wrong number of sequences")
				}

				chromName, chromStart, chromStop, found := parseChromMeta(meta["TR"])
				if !found {
					panic("file format error - missing TR field")
				}

				// TODO: Write helper(s) to check for presence of key and then get the value

				score, err := strconv.Atoi(meta["SC"])
				if err != nil {
					panic("file format error - invalid SC field")
				}

				start, err := strconv.Atoi(meta["ST"])
				if err != nil {
					panic("file format error - invalid ST field")
				}

				stop, err := strconv.Atoi(meta["SP"])
				if err != nil {
					panic("file format error - invalid SP field")
				}

				consensusStart, err := strconv.Atoi(meta["CST"])
				if err != nil {
					panic("file format error - invalid CST field")
				}

				consensusStop, err := strconv.Atoi(meta["CSP"])
				if err != nil {
					panic("file format error - invalid CSP field")
				}

				var sequence string
				var subfamilySequence string
				if meta["TQ"] == "t" {
					sequence = reverse(seqs[0])
					subfamilySequence = reverse(seqs[1])
				} else {
					sequence = seqs[0]
					subfamilySequence = seqs[1]
				}

				flank, err := strconv.Atoi(meta["FL"])
				if err != nil {
					panic("file format error - invalid FL field")
				}

				gapInit, err := strconv.ParseFloat(meta["GI"], 64)
				if err != nil {
					panic("file format error - invalid GI field")
				}

				gapExt, err := strconv.ParseFloat(meta["GE"], 64)
				if err != nil {
					panic("file format error - invalid GE field")
				}

				loaded <- &Alignment{
					Subfamily:         meta["ID"],
					ChromName:         chromName,
					ChromStart:        chromStart,
					ChromStop:         chromStop,
					Score:             score,
					Start:             start,
					Stop:              stop,
					ConsensusStart:    consensusStart,
					ConsensusStop:     consensusStop,
					Sequence:          sequence,
					SubfamilySequence: subfamilySequence,
					Strand:            meta["SD"],
					Flank:             flank,
					SubMatrixName:     meta["MX"],
					GapInit:           gapInit,
					GapExt:            gapExt,
				}

				meta = make(map[string]string)
				seqs = nil
			}

			if !more {
				break
			}
		}

		close(loaded)
	}()

	return loaded
}

func reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}
